Null reset codes are not allowed

mdwheele opened the pull request:
The current schema for Eloquent-based users (maybe others as well, unsure) has  `reset_password_code` is `NULL` by default. The check for correct `$resetCode` simply does an equality comparison and returns true if there is a loose match.

If an attacker is able to provide a null reset code to the package, there are no guards against arbitrary anonymous password resets. In many cases, submitting a url-encoded null byte value (`%00`) will match what's in the database, passing the check and allowing the attacker to set the password to what they wish. Because implementors have reasonable control over the actual field names in their reset forms and processors, there might be some measure of "security through obscurity", but I feel like this is something that the package should cover anyway. Criticality of this is subjective, as a successful attack would need to reset a password and then properly derive the user identity -> email/username. If they have this, they likely have already owned the site. I feel this boils down to a denial of service, otherwise.

I know Sentry is deprecated, but with 1600+ stars and still a good chunk of folks using (https://packagist.org/packages/cartalyst/sentry/stats), I think it's reasonable to mitigate this in some way. My current applications mitigation is to simply disallow "empty" reset codes.

Commit:
Empty reset codes are not allowed

addelindh commented:
I would very much appreciate some indication of whether this patch will be considered at all, or if you are simply not accepting commits because Sentry is deprecated (last commit Nov 26, 2015).

Cheers,
Andreas

drsii commented:
We're considering how best to handle the situation. Accepting means a breaking change and a major version bump.

mdwheele commented:
If I modify to return false rather than throw exception does this resolve the BC issue? I feel like accepting null reset tokens is more of an oversight rather than a feature. Returning false maintains the current contract unless allowing null reset codes was part of that contract to begin with. 

Thoughts?

Commit:
Return false rather than throw exception when reset code is null

mdwheele commented:
@drsii I went ahead and changed the mitigation to return false if an empty reset code is provided. This maintains a boolean return contract while patching the security hole that seems incidental from a schema choice made long ago. I think this is a reasonable compromise and I (as a user of this product, still) would not take issue with this change being in current major releases.

brunogaspar commented:
I don't think you need to use `empty()` here as a simple if check can perform the same and keeps consistency.

``` php
if ( ! $resetCode)
{
    return false;
}
```

Also update the Kohana User class as well please to also return false instead of throwing the exception.

mdwheele commented:
> I don't think you need to use empty() here as a simple if check can perform the same and keeps consistency.

Sounds good. I've modified as suggested and updated `Sentry\Users\Kohana\User` as requested in 006b476.

BackEndTea commented:
Would it be possible to add a release tag for this? 
Currently if i have ```"cartalyst/sentry": "2.1.*"``` in my ```composer.json``` i don't get this security patch.

brunogaspar commented:
@BackEndTea It's tagged.

BackEndTea commented:
Thanks
